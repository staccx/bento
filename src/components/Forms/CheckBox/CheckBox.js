import React from "react"
import PropTypes from "prop-types"
import styled from "styled-components"
import Check from "../../Icons/Check/Check"
import {
  color,
  font,
  spacing,
  fontFamily,
  fontWeight,
  applyVariants,
  targetSize,
  borderRadius,
  commonPropTypes
} from "../../../theming"
import ThemeComponent from "../../Theme/ThemeComponent/ThemeComponent"
import themeProps from "./CheckBox.themeProps"
import { componentCreateFactory } from "../../../theming/utils/createVariantsFunctionFactory"

/**
 * The Checkbox is shown as a square box that is checked when activated.
 Checkboxes are used to let a user select one or more options of a limited number of choices.
 */
const CheckBox = ({
  children,
  id,
  group,
  disabled,
  onChange,
  className,
  defaultChecked,
  checkIcon,
  ignoreBase,
  variant,
  ...otherProps
}) => {
  return (
    <CheckWrapper variant={variant}>
      <InputCheck
        className={className}
        id={id}
        disabled={disabled}
        type="checkbox"
        name={group}
        onChange={onChange}
        defaultChecked={defaultChecked}
        variant={variant}
        {...otherProps}
      />

      <Label htmlFor={id} variant={variant}>
        <Icon variant={variant} />
        {children}
      </Label>
    </CheckWrapper>
  )
}

const CheckWrapper = styled.div`
  ${applyVariants(themeProps.wrapper)};
`

const IconComponent = ({ ...props }) => (
  <ThemeComponent
    tagName={themeProps.iconComponent}
    fallback={Check}
    {...props}
  />
)

export const Icon = styled(IconComponent)`
  position: absolute;
  left: 3px;
  top: 15px;
  display: block;
  height: 18px;
  width: 18px;
  transform: scale(0);
  transition: all 0.2s ease-in-out;
  color: ${color.primary};
  ${applyVariants(themeProps.icon)};
`

const InputCheck = styled.input`
  position: absolute;
  clip: rect(0, 0, 0, 0);
  clip: rect(0 0 0 0);

  &:checked ~ label {
    > svg {
      transform: scale(1);
    }
    ${applyVariants(themeProps.labelChecked)};
  }

  &:checked ~ label::before {
    ${applyVariants(themeProps.labelBeforeChecked)};
  }

  &:focus ~ label {
    &::before {
      border-color: ${color.primary};
    }
  }

  ${applyVariants(themeProps.input)};
`

const Label = styled.label`
  padding: 0 0 0 ${spacing.mediumPlus};
  position: relative;
  font-weight: ${fontWeight.normal};
  letter-spacing: normal;
  width: auto;
  cursor: pointer;
  font-family: ${fontFamily.body};
  font-size: ${font.base};
  line-height: 1.6;
  display: inline-block;
  min-height: ${targetSize.normal};
  padding-top: ${spacing.small};
  padding-bottom: ${spacing.small};

  &::before {
    content: "";
    background: ${color.white};
    border: 1px solid ${color.line};
    border-radius: ${borderRadius};
    display: block;
    height: ${spacing.medium};
    left: 0;
    position: absolute;
    top: ${spacing.small};
    width: ${spacing.medium};
    ${applyVariants(themeProps.labelBefore)};
  }

  &:hover,
  &:focus {
    &::before {
      border-color: ${color.primary};
    }
  }
  ${applyVariants(themeProps.label)};
`

CheckBox.propTypes = {
  children: commonPropTypes.children.isRequired,
  /**
   * If set to true, it cannot be checked
   */
  disabled: PropTypes.bool,
  /**
   * Used to group together multiple checkboxes
   */
  group: PropTypes.string,
  /**
   * Refers to the spesific checkbox, must be unique
   */
  id: PropTypes.string.isRequired,
  input: PropTypes.any,
  /**
   * allows you to preform an action when the component changes
   */
  onChange: PropTypes.func,
  className: PropTypes.string,
  /**
   * if set to true, the component wil render as checked
   */
  defaultChecked: PropTypes.bool,
  /**
   * Allows you to alter the checkmark
   */
  checkIcon: PropTypes.element,
  ignoreBase: PropTypes.func
}

CheckBox.defaultProps = {
  disabled: false,
  input: {},
  onChange: null,
  className: "",
  defaultChecked: false,
  checkIcon: null
}
CheckBox.themeProps = themeProps
CheckBox.createVariants = componentCreateFactory(CheckBox)

export default CheckBox
