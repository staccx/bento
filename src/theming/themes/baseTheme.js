import React from "react"
import Theme from "../Theme"
import reset from "../styles/cssResets"
import grids from "../../components/Layout/Layout/Layout.Grids"
import codeTheme from "../codeTheme"

const Logo = () => <div>ThemeComponent for BaseTheme</div>
Logo.displayName = "Logo for base theme"

export const theme = {
  name: "Base Theme",
  storybookName: "Base Theme",
  spacing: {
    huge: "96px",
    large: "48px",
    mediumPlus: "32px",
    medium: "24px",
    small: "12px",
    tiny: "6px",
    micro: "3px"
  },
  wrapper: {
    small: "420px",
    medium: "640px",
    large: "1024px",
    breakout: "640px"
  },
  color: {
    white: "#fff",
    bg: "#fff",
    subtleHover: "#F7F7FF",
    bgGray: "#f7f7ff",
    line: "#EBEBF2",
    gray: "#767676",
    wcag: "#777",
    disabled: "#BABABA",
    primary: "#00a190",
    secondary: "#F5A623",
    text: "#2b2b2b",
    negative: "#ea526f",
    positive: "#4A90E2",
    warning: "#c54059",
    black: "#000",
    green: "#00a190", // for toggle eks
    link: "#279AF1"
  },
  fontFamily: {
    heading: "sans-serif",
    body: "sans-serif"
  },
  font: {
    h1: "36px",
    h2: "26px",
    h3: "22px",
    h4: "18px",
    h5: "16px",
    h6: "14px",
    base: "14px",
    tiny: "12px",
    input: "16px",
    huge: "44px",
    lineHeight: "1.5"
  },
  fontWeight: {
    bold: "bold",
    normal: "normal",
    light: "lighter"
  },
  targetSize: {
    small: "36px",
    normal: "44px"
  },
  borderRadius: "2px",
  reset,
  grid: grids,
  logo: Logo
}

export const StyleguideTheme = new Theme(theme, codeTheme)

export default theme
