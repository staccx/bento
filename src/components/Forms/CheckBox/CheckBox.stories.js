import React from "react"
import CheckBox from "./CheckBox"

export default {
  title: "New/Components/Forms/CheckBox",
  component: CheckBox
}

export const Checkbox = args => <CheckBox {...args}> Checkbox </CheckBox>
Checkbox.args = { id: "ID" }

export const CheckboxDisabled = args => (
  <CheckBox {...args}> Checkbox </CheckBox>
)
CheckboxDisabled.args = { id: "ID1", disabled: true }

export const CheckboxDefaultChecked = args => (
  <CheckBox {...args}> Checkbox </CheckBox>
)
CheckboxDefaultChecked.args = { id: "ID2", defaultChecked: true }

export const CheckboxConsoleLog = args => (
  <CheckBox {...args}> Checkbox </CheckBox>
)
CheckboxConsoleLog.args = {
  id: "ID3",
  onChange: e => console.log("test", e.target.value, e.target.id)
}

export const Multiple = () => {
  return (
    <>
      <CheckBox id="1ID" group="test">
        Checkbox 1
      </CheckBox>
      <CheckBox id="2ID" group="test">
        Checkbox 2
      </CheckBox>
    </>
  )
}

export const SameId = () => {
  return (
    <>
      <CheckBox id="1">Checkbox 1 </CheckBox>
      <CheckBox id="1">Checkbox 2 </CheckBox>
    </>
  )
}
