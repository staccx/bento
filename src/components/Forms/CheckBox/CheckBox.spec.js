import React from "react"
import renderer from "react-test-renderer"
import { ThemeProvider } from "styled-components"
import baseTheme from "../../../theming/themes/baseTheme"
import { render, fireEvent, screen } from "@testing-library/react"
import CheckBox from "./CheckBox"

const Child = () => <p>Body</p>

describe("CheckBox", () => {
  describe("Snapshots", () => {
    it("No props", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <CheckBox>
              <Child />
            </CheckBox>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("with Id", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <CheckBox id="ID">
              <Child />
            </CheckBox>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("onChange", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <CheckBox id="ID" onChange={() => console.log("changed!")}>
              <Child />
            </CheckBox>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("group", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <CheckBox id="ID" group="124">
              <Child />
            </CheckBox>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})

describe("Rendering", () => {
  const tempConsole = console.log
  beforeAll(() => {
    console.log = jest.fn()
  })
  afterAll(() => {
    console.log = tempConsole
  })
  it("Should click", () => {
    const onChangeMock = jest.fn()
    render(
      <ThemeProvider theme={baseTheme}>
        <CheckBox
          data-testid="test"
          id="ID"
          value="test"
          onChange={e => onChangeMock(e.target.value)}
        >
          <Child />
        </CheckBox>
      </ThemeProvider>
    )
    const checkbox = screen.getByTestId("test")
    expect(checkbox).toBeInTheDocument()
    expect(checkbox.checked).toBe(false)
    fireEvent.click(checkbox)
    expect(onChangeMock).toBeCalledWith("test")
    expect(checkbox.checked).toBe(true)
  })

  it("DefaultChecked", () => {
    const onChangeMock = jest.fn()
    render(
      <ThemeProvider theme={baseTheme}>
        <CheckBox
          data-testid="test"
          id="ID"
          defaultChecked
          onChange={onChangeMock}
        >
          <Child />
        </CheckBox>
      </ThemeProvider>
    )
    const checkbox = screen.getByTestId("test")
    expect(checkbox).toBeInTheDocument()
    expect(checkbox.checked).toBe(true)
    fireEvent.click(checkbox)
    expect(onChangeMock).toBeCalledTimes(1)
    expect(checkbox.checked).toBe(false)
  })
})
