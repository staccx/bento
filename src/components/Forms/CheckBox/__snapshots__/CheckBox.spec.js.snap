// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBox Snapshots No props 1`] = `
<div
  className="sc-gsTCUz gBEZiA"
>
  <input
    className="sc-hKgILt giIjLZ"
    defaultChecked={false}
    disabled={false}
    onChange={null}
    type="checkbox"
  />
  <label
    className="sc-eCssSg kpsaBE"
  >
    <svg
      className="sc-bdfBwQ hZduwO sc-dlfnbm jaGLwg"
      viewBox="0 0 24 24"
    >
      <path
        d="M23.43 1.75L21.27.23c-.6-.41-1.4-.25-1.8.36L8.9 16.72l-4.84-5.03a1.27 1.27 0 0 0-1.83 0l-1.85 1.9a1.4 1.4 0 0 0 0 1.92l7.46 7.73A2.5 2.5 0 0 0 9.5 24c.6 0 1.2-.37 1.57-.96L23.78 3.63a1.4 1.4 0 0 0-.36-1.88h.01z"
      />
    </svg>
    <p>
      Body
    </p>
  </label>
</div>
`;

exports[`CheckBox Snapshots group 1`] = `
<div
  className="sc-gsTCUz gBEZiA"
>
  <input
    className="sc-hKgILt giIjLZ"
    defaultChecked={false}
    disabled={false}
    id="ID"
    name="124"
    onChange={null}
    type="checkbox"
  />
  <label
    className="sc-eCssSg kpsaBE"
    htmlFor="ID"
  >
    <svg
      className="sc-bdfBwQ hZduwO sc-dlfnbm jaGLwg"
      viewBox="0 0 24 24"
    >
      <path
        d="M23.43 1.75L21.27.23c-.6-.41-1.4-.25-1.8.36L8.9 16.72l-4.84-5.03a1.27 1.27 0 0 0-1.83 0l-1.85 1.9a1.4 1.4 0 0 0 0 1.92l7.46 7.73A2.5 2.5 0 0 0 9.5 24c.6 0 1.2-.37 1.57-.96L23.78 3.63a1.4 1.4 0 0 0-.36-1.88h.01z"
      />
    </svg>
    <p>
      Body
    </p>
  </label>
</div>
`;

exports[`CheckBox Snapshots onChange 1`] = `
<div
  className="sc-gsTCUz gBEZiA"
>
  <input
    className="sc-hKgILt giIjLZ"
    defaultChecked={false}
    disabled={false}
    id="ID"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="sc-eCssSg kpsaBE"
    htmlFor="ID"
  >
    <svg
      className="sc-bdfBwQ hZduwO sc-dlfnbm jaGLwg"
      viewBox="0 0 24 24"
    >
      <path
        d="M23.43 1.75L21.27.23c-.6-.41-1.4-.25-1.8.36L8.9 16.72l-4.84-5.03a1.27 1.27 0 0 0-1.83 0l-1.85 1.9a1.4 1.4 0 0 0 0 1.92l7.46 7.73A2.5 2.5 0 0 0 9.5 24c.6 0 1.2-.37 1.57-.96L23.78 3.63a1.4 1.4 0 0 0-.36-1.88h.01z"
      />
    </svg>
    <p>
      Body
    </p>
  </label>
</div>
`;

exports[`CheckBox Snapshots with Id 1`] = `
<div
  className="sc-gsTCUz gBEZiA"
>
  <input
    className="sc-hKgILt giIjLZ"
    defaultChecked={false}
    disabled={false}
    id="ID"
    onChange={null}
    type="checkbox"
  />
  <label
    className="sc-eCssSg kpsaBE"
    htmlFor="ID"
  >
    <svg
      className="sc-bdfBwQ hZduwO sc-dlfnbm jaGLwg"
      viewBox="0 0 24 24"
    >
      <path
        d="M23.43 1.75L21.27.23c-.6-.41-1.4-.25-1.8.36L8.9 16.72l-4.84-5.03a1.27 1.27 0 0 0-1.83 0l-1.85 1.9a1.4 1.4 0 0 0 0 1.92l7.46 7.73A2.5 2.5 0 0 0 9.5 24c.6 0 1.2-.37 1.57-.96L23.78 3.63a1.4 1.4 0 0 0-.36-1.88h.01z"
      />
    </svg>
    <p>
      Body
    </p>
  </label>
</div>
`;
